core/src/test/scala/unit/kafka/server/AddPartitionsToTxnManagerTest.scala

```diff
@@ -68,87 +71,143 @@ class AddPartitionsToTxnManagerTest {
       if (requestAndHandler.destination == node0) {
         assertEquals(time.milliseconds(), requestAndHandler.creationTimeMs)
-        assertEquals(AddPartitionsToTxnRequest.Builder.forBroker(
-          new AddPartitionsToTxnTransactionCollection(Seq(transactionData(transactionalId3, producerId3), transactionData(transactionalId1, producerId1, producerEpoch = 1)).iterator.asJava)).data,
-          requestAndHandler.request.asInstanceOf[AddPartitionsToTxnRequest.Builder].data) // insertion order
+        assertEquals(
+          AddPartitionsToTxnRequest.Builder.forBroker(
+            new AddPartitionsToTxnTransactionCollection(Seq(
+              transactionData(transactionalId3, producerId3, verifyOnly = true),
+              transactionData(transactionalId1, producerId1, producerEpoch = 1, verifyOnly = true)
+            ).iterator.asJava)
+          ).data,
+          requestAndHandler.request.asInstanceOf[AddPartitionsToTxnRequest.Builder].data // insertion order
+        )
       } else {
-        verifyRequest(node1, transactionalId2, producerId2, requestAndHandler)
+        verifyRequest(node1, transactionalId2, producerId2, requestAndHandler, verifyOnly = true)
       }
     }
   }
 
   @Test
   def testGenerateRequests(): Unit = {
+    when(partitionFor.apply(transactionalId1)).thenReturn(0)
+    when(partitionFor.apply(transactionalId2)).thenReturn(1)
+    when(partitionFor.apply(transactionalId3)).thenReturn(2)
+    when(metadataCache.getTopicMetadata(Set(Topic.TRANSACTION_STATE_TOPIC_NAME), config.interBrokerListenerName))
+      .thenReturn(Seq(
+        new MetadataResponseData.MetadataResponseTopic()
+          .setName(Topic.TRANSACTION_STATE_TOPIC_NAME)
+          .setPartitions(List(
+            new MetadataResponseData.MetadataResponsePartition()
+              .setPartitionIndex(0)
+              .setLeaderId(0),
+            new MetadataResponseData.MetadataResponsePartition()
+              .setPartitionIndex(1)
+              .setLeaderId(1),
+            new MetadataResponseData.MetadataResponsePartition()
+              .setPartitionIndex(2)
+              .setLeaderId(2)
+          ).asJava)
+      ))
+    when(metadataCache.getAliveBrokerNode(0, config.interBrokerListenerName))
+      .thenReturn(Some(node0))
+    when(metadataCache.getAliveBrokerNode(1, config.interBrokerListenerName))
+      .thenReturn(Some(node1))
+    when(metadataCache.getAliveBrokerNode(2, config.interBrokerListenerName))
+      .thenReturn(Some(node2))
+
     val transactionErrors = mutable.Map[TopicPartition, Errors]()
 
-    addPartitionsToTxnManager.addTxnData(node0, transactionData(transactionalId1, producerId1), setErrors(transactionErrors))
-    addPartitionsToTxnManager.addTxnData(node1, transactionData(transactionalId2, producerId2), setErrors(transactionErrors))
+    addPartitionsToTxnManager.addTxnData(transactionalId1, producerId1, producerEpoch = 0, verifyOnly = true, topicPartitions, setErrors(transactionErrors))
```

Comment:
See my previous comment about verifyOnly = false. I suppose this is helpful for setting up for part 2 though :)